### PR TITLE
Make buffer of scraped particles available to Python code

### PR DESCRIPTION
**Pull request by PhilMiller**
Based on Andrew's @atmyers  PR #2124

Work by @MKieburtz and @roelof-groenewald 

- [x] rebase after #2124 was merged
- [x] add documentation of scraping parameters from https://github.com/ECP-WarpX/WarpX/pull/2124#discussion_r693154853 :)

### CODE REVIEW

File: Examples/Modules/ParticleBoundaryScrape/PICMI_inputs_wrapper_scrape.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+sim.add_diagnostic(field_diag)
+
+sim.step(max_steps - 1)
+
+# check scraped particle buffer part way through the simulation
+from pywarpx import _libwarpx
+
+n = _libwarpx.get_num_particles_impacted_boundary("electrons", 'x_hi')
+print("Number of electrons scraped:", n)
+weights = _libwarpx.get_particle_boundary_buffer("electrons", 'x_hi', 'w', 0)
```
Comment (atmyers) [inline Examples/Modules/ParticleBoundaryScrape/PICMI_inputs_wrapper_scrape.py]: Would it be possible / desirable to set this up in such a way that it computes the total charge absorbed at each time step, and compares it to what it should be based on the simulation setup?

Comment (roelof-groenewald) [reply Examples/Modules/ParticleBoundaryScrape/PICMI_inputs_wrapper_scrape.py]: I guess the best thing to do would be to make this test match yours exactly and then just also run the commands to check that the _libwarpx wrappers function properly. That way we can even use the same analysis file.

**Review by atmyers — APPROVED** [review]

**Comment by PhilMiller** [comment]
(converted to draft until Michael and Roelof can check that my rebase was successful, and re-activate the Python test)

File: Docs/source/usage/parameters.rst [hunk 2]
```diff
@@ -759,6 +759,26 @@ Particle initialization
...
     by any pusher during the simulation.
 
+* ``<species>.save_particles_at_xlo/ylo/zlo``,  ``<species>.save_particles_at_xhi/yhi/zhi`` and ``<species>.save_particles_at_eb`` (`0` or `1` optional, default `0`)
+    If `1` particles of this species will be copied to the scraped particle
+    buffer for the specified boundary if they leave the simulation domain in
+    the specified direction. **If USE_EB=TRUE** the ``save_particles_at_eb``
+    flag can be set to `1` to also save particle data for the particles of this
+    species that impact the embedded boundary.
+    The scraped particle buffer can be used to track particle fluxes out of the
+    simulation. If Python is used to run the simulation ``pywarpx._libwarpx``
```
Comment (dpgrote) [inline Docs/source/usage/parameters.rst]: Can a comment be added here that the particle buffers are currently only accessible via the Python interface?

File: Docs/source/usage/parameters.rst [hunk 3]
```diff
@@ -759,6 +759,26 @@ Particle initialization
...
+    function ``get_particle_boundary_buffer()`` can be used to access the
+    scraped particle buffer. The scraped particle buffer includes an entry for
+    every particle of the timestep at which the particle was scraped. This can
+    be accessed by passing the argument ``comp_name="step_scraped"`` to the above
+    mentioned function.
+
+    .. note::
+        Currently the scraped particle buffer relies on the user to access the
+        data in the buffer for processing and periodically clear the buffer. The
```
Comment (dpgrote) [inline Docs/source/usage/parameters.rst:778]: Suggestion "...clear the buffer by calling warpx_clearParticleBoundaryBuffer".

File: Python/pywarpx/_libwarpx.py [hunk 4]
```diff
@@ -763,6 +767,122 @@ def add_real_comp(species_name, pid_name, comm=True):
...
+        dimensions = {'x' : 0, 'y' : 1, 'z' : 2}
+    elif geometry_dim == '2d':
+        dimensions = {'x' : 0, 'z' : 1}
+    else:
+        raise NotImplementedError("RZ is not supported for particle scraping.")
+
+    if boundary != 'eb':
+        boundary_parts = boundary.split("_")
+        dim_num = dimensions[boundary_parts[0]]
+        side = 0 if boundary_parts[1] == 'lo' else 1
```
Comment (dpgrote) [inline Python/pywarpx/_libwarpx.py]: This doesn't check if boundary_parts[1] is `hi`. The input could be `z_foo` and this wouldn't complain, returning the same thing as if it were `z_hi`.

File: Source/Evolve/WarpXEvolve.cpp [hunk 5]
```diff
@@ -295,6 +295,8 @@ WarpX::Evolve (int numsteps)
             mypc->SortParticlesByBin(sort_bin_size);
         }
 
+        if (warpx_py_beforeEsolve) warpx_py_beforeEsolve();
```
Comment (dpgrote) [inline Source/Evolve/WarpXEvolve.cpp]: Is this change intended? It is not directly related to the main topic. Also, this repeats the call to `warpx_py_beforeEsolve`, as it is also called below. This would be better in a separate PR.

Comment (PhilMiller) [reply Source/Evolve/WarpXEvolve.cpp]: That was probably a rebase artifact, relative to some re-ordering in here.

File: Source/Particles/ParticleBoundaryBuffer.H [hunk 6]
```diff
@@ -33,13 +33,24 @@ public:
         return m_species_names;
     }
 
+    int getSpeciesIndexFromName ( const std::string name ) const {
+        auto it = std::find(m_species_names.begin(), m_species_names.end(), name);
```
Comment (dpgrote) [inline Source/Particles/ParticleBoundaryBuffer.H]: There is already the routine `MultiParticleContainer::getSpeciesID`. Can that be used here, simply returning its result?

**Review by dpgrote — COMMENTED** [review]
Overall this looks good. I have several minor comments.

File: Source/Evolve/WarpXEvolve.cpp (hunk 5 above)
Comment (roelof-groenewald) [reply Source/Evolve/WarpXEvolve.cpp]: I've removed it here and will submit a follow-up PR for it. The issue is that currently `warpx_py_beforeEsolve()` is only called when using the electromagnetic solver, so I'll add a call for it inside the electrostatic solver block.

File: Source/Particles/ParticleBoundaryBuffer.H (hunk 6 above)
Comment (roelof-groenewald) [reply Source/Particles/ParticleBoundaryBuffer.H]: Thanks for pointing that out. I wasn't aware of the `getSpeciesID` function so it is good to know about it now.

**Review by dpgrote — APPROVED** [review]
Thanks for the changes! This looks ready to merge.

File: Docs/source/usage/parameters.rst [hunk 7]
```diff
@@ -759,6 +759,27 @@ Particle initialization
     If `1` is given, this species will not be pushed
     by any pusher during the simulation.
 
+* ``<species>.save_particles_at_xlo/ylo/zlo``,  ``<species>.save_particles_at_xhi/yhi/zhi`` and ``<species>.save_particles_at_eb`` (`0` or `1` optional, default `0`)
+    If `1` particles of this species will be copied to the scraped particle
+    buffer for the specified boundary if they leave the simulation domain in
+    the specified direction. **If USE_EB=TRUE** the ``save_particles_at_eb``
```
Comment (ax3l) [inline Docs/source/usage/parameters.rst:765]: We usually document the CMake flags now:
```suggestion
    the specified direction. **If WarpX_EB=ON** the ``save_particles_at_eb``
```

**Review by ax3l — COMMENTED** [review]
Will merge now, little documentation hit for follow-ups inline.
